### ISSUE TITLE
[Explicit] tests are executed in Rider/R#

### ISSUE THREAD

**Issue by roji** [nunit/nunit3-vs-adapter#487]
When running the tests for my entire project in Rider, tests marked with `[Explicit]` are executed. [This issue was investigated by Jetbrains](https://youtrack.jetbrains.com/issue/RIDER-11989), who concluded that it's an issue on the NUnit side, where there adapter is hard-coded to run explicit ones. The issue also occurs on Visual Studio 15.6.2 with Resharper 2017.3.3.

NUnit: 3.10.1
NUnit3TestAdapter: 3.10.0
Rider: 2017.3.1 (on Linux)
Targets: net45, netcoreapp2.0

Test project csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
  <PropertyGroup>
    <TargetFrameworks>net451;netcoreapp2.0</TargetFrameworks>
    <DebugType>portable</DebugType>
    <AssemblyOriginatorKeyFile>../../Npgsql.snk</AssemblyOriginatorKeyFile>
    <SignAssembly>true</SignAssembly>
    <PublicSign Condition=" '$(OS)' != 'Windows_NT' ">true</PublicSign>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="../../src/Npgsql/Npgsql.csproj" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="NUnit" Version="3.10.1" />
    <PackageReference Include="NLog" Version="5.0.0-beta06" />
    <PackageReference Include="Microsoft.CSharp" Version="4.4.1" />
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.5.0" />
    <PackageReference Include="NUnit3TestAdapter" Version="3.10.0" />
  </ItemGroup>
  <ItemGroup Condition=" '$(TargetFramework)' == 'net451' ">
    <Reference Include="System.Transactions" />
  </ItemGroup>
  <ItemGroup>
    <Reference Include="Microsoft.CSharp" />
  </ItemGroup>
</Project>
```

**Comment by roji** [comment]
Note: in the old runner there was https://github.com/nunit/dotnet-test-nunit/issues/25, which led to https://github.com/nunit/nunit3-vs-adapter/issues/47, but that issue is marked as resolved in 3.10, and that doesn't seem to be the case.

**Comment by OsirisTerje** [comment]
The NUnit3TestAdapter is not used by JetBrains, they have their own adapter.

**Comment by rprouse** [comment]
The old test adapter only worked for .NET Core 1.0 and doesn't share code with the other adapters. If JetBrains is somehow using our adapter, it is likely that they are sending us filters for running all tests (as opposed to an empty filter) which explicitly tells us to run all tests which we take to mean including explicit. We'll need more information from the JetBrain's team on how they use our adapter and which filters they pass for `Run All Tests`.

**Comment by roji** [comment]
@rprouse I hope you guys can work with JetBrains to get this resolved...

**Comment by jnm2** [comment]
There are two entry through which VSTest tells adapters like ours to run tests:

 - An entry point which hands us a parsed filter expression. This is invoked through the CLI.

 - An entry point which hands us a precise list of previously discovered _test cases_ to execute, some or all of which may be explicit. This is invoked when you run a selection of tests from Test Explorer. Test Explorer gives us a selection to run without telling how the seelction was made, so we can't know if you explicitly selected each test case or if it was thrown in under a group header. (@OsirisTerje, is it easy for you to share your screenshots which demonstrate this?)
   
   The only way to truly fix this is if Test Explorer starts actually understanding the concept of explicit selection, or if Test Explorer starts telling us what the UI element was that the user actually clicked.

I'm going to assume Rider is using the latter entry point. If so, the same things that I stated about VS's Test Explorer apply equally to Rider.

@OsirisTerje brought up the idea again that for the latter entry point, we could skip all explicit tests unless the selection comprises *only* explicit tests. I think that is the absolute best option, given the inherent constraints of the latter entry point.

**Comment by estrizhok** [comment]
@rprouse in JetBrains we do use your adapter to run tests for .netcore, and yes you are right that we send to the adapter list of tests that needed to be executed. that list might include tests marked with [Explicit] attribute, if user had those in their selection. That selection is the reason why send a list populated with specific tests instead of an empty list. We could have easily excluded tests marked with [Explicit] from the run, if we such information. Unfortunately, test-cases returned by the adapter on discovery step do not have . Of course, we can run our own analysis, then merge its result with what test adapter returns, but we would rather use the information provided by the adapter only. 
In this regard, could you add an additional custom property to test-cases indicating whether when they are explicit?

**Comment by OsirisTerje** [comment]
From version 3.10 of the adapter, all test cases that have an Explicit attribute also has a category added with the same name.

**Comment by estrizhok** [comment]
I have just checked the latest version of adapter (3.10), and I see that Explicit attribute not only passed as a category, but also as a custom property. Excellent!

**Comment by OsirisTerje** [comment]
Thanks!  Superb!  Then I hope it works out!

**Comment by jnm2** [comment]
🎉 This is great news!

Besides the user-visible Explicit category, the hidden bool [NUnit.Explicit property](https://github.com/nunit/nunit3-vs-adapter/pull/471/files#diff-080906f049778bf6c6789702b9a89a35R16) which @estrizhok is talking about was added in https://github.com/nunit/nunit3-vs-adapter/pull/471. We had to do this in order to know in the execution phase which ones were  truly explicit in the discovery phase.

Now that Rider will be using the NUnit.Explicit bool property as an augmenting side channel, we should add tests which ensure that (for example) this hidden property keeps existing with the same name and type. Even if we end up not using it internally, we still want Rider to be able to rely on it.

@estrizhok Will ReSharper gain the same understanding of explicit selections as Rider?
